File: julia/matrix_chain.jl
function matrix_chain(n)

  reps = parse(Int64, ENV["LAMP_REPS"])
  m = Int(floor(n / 5))

  A = randn(n,n)
  B = randn(n,n)
  C = randn(m,n)

  Benchmarker.add_data(csv, "mc_l_r_orig", Benchmarker.measure(reps, left_to_right_chain_original, A, B, C))
  Benchmarker.add_data(csv, "mc_l_r_guid", Benchmarker.measure(reps, left_to_right_chain_guided, A, B, C))

  @test isapprox(left_to_right_chain_original(A, B, C)[2], left_to_right_chain_guided(A, B, C)[2])

  C = randn(n,m)
  Benchmarker.add_data(csv, "mc_r_l_orig", Benchmarker.measure(reps, right_to_left_chain_original, A, B, C))
  Benchmarker.add_data(csv, "mc_r_l_guid", Benchmarker.measure(reps, right_to_left_chain_guided, A, B, C))

  @test isapprox(right_to_left_chain_original(A, B, C)[2], right_to_left_chain_guided(A, B, C)[2])

  A = randn(n,n)
  B = randn(n,m)
  C = randn(m,n)
  D = randn(n,n)

  Benchmarker.add_data(csv, "mc_mixed_orig", Benchmarker.measure(reps, mixed_chain_original, A, B, C, D))
  Benchmarker.add_data(csv, "mc_mixed_guid", Benchmarker.measure(reps, mixed_chain_guided, A, B, C, D))

  @test isapprox(mixed_chain_original(A, B, C, D)[2], mixed_chain_guided(A, B, C, D)[2])
end

function left_to_right_chain_original(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2})
  result = zeros(size(C,1), size(B,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = C * A * B

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

function left_to_right_chain_guided(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2})
  result = zeros(size(C,1), size(B,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = (C * A) * B

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

function right_to_left_chain_original(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2})
  result = zeros(size(A,1), size(C,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = A * B * C

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

function right_to_left_chain_guided(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2})
  result = zeros(size(A,1), size(C,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = A * (B * C)

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

function mixed_chain_original(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2}, D::DenseArray{Float64, 2})
  result = zeros(size(A,1), size(D,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = A * B * C * D

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

function mixed_chain_guided(A::DenseArray{Float64, 2}, B::DenseArray{Float64, 2}, C::DenseArray{Float64, 2}, D::DenseArray{Float64, 2})
  result = zeros(size(A,1), size(D,2))
  start::Float64 = 0.0
  finish::Float64 = 0.0
  Benchmarker.cachescrub()
  Benchmarker.gcscrub()
  start = time_ns()

  result = (A * B) * (C * D)

  finish = time_ns()
  return ((finish-start)*1e-9, result)
end

File: matlab/matrix_chain.m
function matrix_chain(A, B, b)

reps = str2num(getenv('LAMP_REPS'));

n = size(A, 2);

C1 = randn(n/5, n);
C2 = randn(n, n/5);
l_r_orig_ = @() l_r_orig(A, B, C1);
l_r_guid_ = @() l_r_guid(A, B, C1);
r_l_orig_ = @() r_l_orig(A, B, C2);
r_l_guid_ = @() r_l_guid(A, B, C2);

C11 = randn(n, n/5);
C21 = randn(n/5, n);
mixed_orig_ = @() mixed_orig(A, B, C11, C21);
mixed_guid_ = @() mixed_guid(A, B, C11, C21);

b.benchmark('mc_l_r_orig', reps, l_r_orig_);
b.benchmark('mc_l_r_guid', reps, l_r_guid_);
b.benchmark('mc_r_l_orig', reps, r_l_orig_);
b.benchmark('mc_r_l_guid', reps, r_l_guid_);
b.benchmark('mc_mixed_orig', reps, mixed_orig_);
b.benchmark('mc_mixed_guid', reps, mixed_guid_);
end

function [C, time] = l_r_orig(A, B, m)
C = zeros(size(m, 1), size(B, 2));
tic;
C = m * A * B;
time = toc;
end

function [C, time] = l_r_guid(A, B, m)
C = zeros(size(m, 1), size(B, 2));
tic;
C = (m * A) * B;
time = toc;
end

function [C, time] = r_l_orig(A, B, m)
C = zeros(size(A, 1), size(m, 2));
tic;
C = A * B * m;
time = toc;
end

function [C, time] = r_l_guid(A, B, m)
C = zeros(size(A, 1), size(m, 2));
tic;
C = A * (B * m);
time = toc;
end

function [C, time] = mixed_orig(A, B, m1, m2)
C = zeros(size(A, 1), size(B, 2));
tic;
C = A * m1 * m2 * B;
time = toc;
end

function [C, time] = mixed_guid(A, B, m1, m2)
C = zeros(size(A, 1), size(B, 2));
tic;
C = (A * m1) * (m2 * B);
time = toc;
end
